File: 2019/02/rs/src/main.rs
use std::env;
use std::fs;
use std::time::{Instant};
use std::cell::RefCell;

pub struct IntCodeComputer {
    memory: RefCell<Vec<i32>>,
    pointer: usize,
    running: bool,
}

impl IntCodeComputer {
    pub fn new(memory: &Vec<i32>) -> IntCodeComputer {
        IntCodeComputer {
            memory: RefCell::new(memory.clone()),
            pointer: 0,
            running: true
        }
    }

    pub fn run(&mut self) -> i32 {
        while self.running {
            self.tick();
        }
        self.memory.borrow()[0]
    }

    fn get_parameter(&self, offset: i32) -> i32 {
        self.memory.borrow()[self.memory.borrow()[self.pointer + offset as usize] as usize]
    }

    fn get_address(&self, offset: i32) -> usize {
        self.memory.borrow()[self.pointer + offset as usize] as usize
    }

    fn tick(&mut self) {
        if !self.running {
            return
        }
        let opcode = self.memory.borrow()[self.pointer];
        match opcode {
            1 => { // ADD
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] = self.get_parameter(1) + self.get_parameter(2);
                self.pointer += 4;
            },
            2 => { // MUL
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] = self.get_parameter(1) * self.get_parameter(2);
                self.pointer += 4;
            },
            99 => { // HALT
                self.running = false;  
            }
            _ => panic!("Unknown instruction {} {}", self.pointer, opcode)
        }
    }
}

fn run_program(memory: &Vec<i32>, noun: i32, verb: i32) -> i32 {
    let mut memory = memory.clone();
    memory[1] = noun;
    memory[2] = verb;
    IntCodeComputer::new(&memory).run()
}

const TARGET_VALUE: i32 = 19690720;
fn part2(memory: &Vec<i32>) -> i32 {
    for noun in 0..100 {
        for verb in 0..100 {
            if run_program(memory, noun, verb) == TARGET_VALUE {
                return 100 * noun + verb
            }
        }
    };
    panic!("Target value not found")
}

fn solve(memory: &Vec<i32>) -> (i32, i32) {
    (run_program(&memory, 12, 2), part2(memory))
}

fn get_input(file_path: &String) -> Vec<i32> {
    fs::read_to_string(file_path).expect("Error reading input file!")
        .split(',')
        .map(|i| i.trim().parse().unwrap())
        .collect()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:?}", now.elapsed().as_secs_f32());
}

File: 2019/02/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <string.h>

typedef struct
{
    int *memory;
    size_t length;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

typedef struct
{
    int *memory;
    int pointer;
    int running;
} IntCodeComputer;

int getParameter(IntCodeComputer *computer, int offset)
{
    return computer->memory[computer->memory[computer->pointer + offset]];
}

int getAddress(IntCodeComputer *computer, int offset)
{
    return computer->memory[computer->pointer + offset];
}

char errorMessage[64];
void tick(IntCodeComputer *computer)
{
    if (!computer->running)
        return;
    int opcode = computer->memory[computer->pointer];
    switch (opcode)
    {
    case 1: // ADD
        computer->memory[getAddress(computer, 3)] = getParameter(computer, 1) + getParameter(computer, 2);
        computer->pointer += 4;
        break;
    case 2: // MUL
        computer->memory[getAddress(computer, 3)] = getParameter(computer, 1) * getParameter(computer, 2);
        computer->pointer += 4;
        break;
    case 99: // HALT
        computer->running = 0;
        break;
    default:
        sprintf(errorMessage, "Unknow instruction '%d' at '%d'\n", opcode, computer->pointer);
        perror(errorMessage);
        exit(1);
        break;
    }
}

IntCodeComputer newComputer(Input input)
{
    int *newMemory = calloc(input.length, sizeof(int));
    int *address = newMemory;
    while (input.length--)
        *(address++) = *(input.memory++);
    return (IntCodeComputer){
        newMemory,
        0,
        1};
}

int run(IntCodeComputer *computer)
{
    while (computer->running)
        tick(computer);
    return computer->memory[0];
}

int runProgram(Input input, int noun, int verb)
{
    IntCodeComputer computer = newComputer(input);
    computer.memory[1] = noun;
    computer.memory[2] = verb;
    return run(&computer);
}

const int TARGET_VALUE = 19690720;

int part2(Input input)
{
    int noun, verb;
    for (noun = 0; noun < 100; noun++)
        for (verb = 0; verb < 100; verb++)
            if (runProgram(input, noun, verb) == TARGET_VALUE)
                return 100 * noun + verb;
    perror("Target value not found");
    exit(1);
}

Results solve(Input input)
{
    return (Results){runProgram(input, 12, 2), part2(input)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    char *content = malloc(length);
    fread(content, 1, length, file);
    int *memory = calloc(512, sizeof(int));
    int count = 0;
    char *value = strtok(content, ",");
    while (value != NULL)
    {
        memory[count++] = atoi(value);
        value = strtok(NULL, ",");
    }
    fclose(file);
    return (Input){memory, count};
}

void freeInput(Input input)
{
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}